### PR TITLE
Show experience gain rate

### PR DESCRIPTION
**Pull request by omarcopires**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed 
Now send rate as expected instead of always sending a fixed value to the player.

image

**Note:** I tried to send with LUA, but I didn't get anything but a crash, if someone wants to teach me or show me how it can be done I would be grateful.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: data/creaturescripts/scripts/login.lua [hunk 1]
```diff
@@ -22,6 +22,9 @@ function onLogin(player)
 		player:setVocation(vocation:getDemotion())
 	end
 
+	-- Update client xp gain rate
+	player:updateClientXpGainRate()
```
Comment (ranisalt) [inline data/creaturescripts/scripts/login.lua:26]: Experience is always abbreviated to `exp`, not `xp`, in the sources

File: data/scripts/creaturescripts/monster/gs_wyda_death.lua
Comment (ranisalt) [inline data/scripts/creaturescripts/monster/gs_wyda_death.lua:1]: These files seem unrelated to the change and should not have been staged

Comment (omarcopires) [reply data/scripts/creaturescripts/monster/gs_wyda_death.lua:1]: it's just to improve the organization, but I'll get back to them

File: config.lua.dist [hunk 2]
```diff
@@ -98,6 +99,7 @@ forceMonsterTypesOnLoad = true
 cleanProtectionZones = false
 showPlayerLogInConsole = true
 checkDuplicateStorageKeys = false
+expDisplayRates = true
```
Comment (soul4soul) [inline config.lua.dist:102]: Is a configuration value really required for this feature?

Comment (ranisalt) [reply config.lua.dist:102]: I don't think so, should always be enabled

File: data/lib/core/player.lua [hunk 3]
```diff
@@ -536,23 +536,17 @@ function Player.sendBestiaryMilestoneReached(self, raceId)
...
-	-- Experience display
-	local xpGainRate = 0
-	if configManager.getBoolean(configKeys.EXP_DISPLAY_MODE) then
-		xpGainRate = 100 * Game.getExperienceStage(self:getLevel())
-	else
-		xpGainRate = 100
-	end
-
+	-- Experience bonus (includes server rates)
+	local xpGainRate = 100 * Game.getExperienceStage(self:getLevel())
```
Comment (ranisalt) [inline data/lib/core/player.lua:540]:
```suggestion
	local expGainRate = 100 * Game.getExperienceStage(self:getLevel())
```

File: data/lib/core/player.lua [hunk 4]
```diff
@@ -534,3 +534,21 @@ function Player.sendBestiaryMilestoneReached(self, raceId)
...
+	-- Stamina bonus
+	local staminaMinutes = self:getStamina()
+	local staminaBonus = (staminaMinutes > 2340) and 150 or ((staminaMinutes < 840) and 50 or 100)
+	self:setStaminaExpBoost(staminaBonus)
+
+	-- Low level grinding bonus
+	-- TODO: Implement bonus system for low levels
+	local level = self:getLevel()
+	local levelBonus = (level > 1 and level <= 50) and 50 or 0
+	self:setGrindingExpBoost(levelBonus)
```
Comment (ranisalt) [inline data/lib/core/player.lua:552]: Where does the name "grinding exp" come from?

Comment (omarcopires) [reply data/lib/core/player.lua:552]: this bonus is the one with a level lower than 50, I couldn't implement it and I don't even have a better name for it, if you want to suggest something.

https://tibia.fandom.com/wiki/Experience_Point
image
